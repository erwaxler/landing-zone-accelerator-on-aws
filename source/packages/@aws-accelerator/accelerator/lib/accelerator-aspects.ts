/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { IConstruct } from 'constructs';
import { version } from '../../../../package.json';

/**
 * Property overrides for GovCloud environments
 */
class GovCloudOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = (node as cdk.aws_iam.CfnRole).assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc.toJSON()).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.amazonaws-us-gov.com/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO-B environments
 */
class IsobOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_ec2.CfnFlowLog) {
      node.addPropertyDeletionOverride('LogFormat');
      node.addPropertyDeletionOverride('Tags');
      node.addPropertyDeletionOverride('MaxAggregationInterval');
    }
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_s3.CfnBucket) {
      node.addPropertyDeletionOverride('PublicAccessBlockConfiguration');
      node.addPropertyDeletionOverride('OwnershipControls');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('InsightSelectors');
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
    if (node instanceof cdk.aws_ec2.CfnVPCEndpoint) {
      const ServiceName = node.serviceName.replace('com.amazonaws.us', 'gov.sgov.sc2s.us');
      node.addPropertyOverride('ServiceName', ServiceName);
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = (node as cdk.aws_iam.CfnRole).assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc.toJSON()).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.sc2shome.sgov.gov/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO environments
 */
class IsoOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_ec2.CfnFlowLog) {
      node.addPropertyDeletionOverride('LogFormat');
      node.addPropertyDeletionOverride('Tags');
      node.addPropertyDeletionOverride('MaxAggregationInterval');
    }
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_s3.CfnBucket) {
      node.addPropertyDeletionOverride('PublicAccessBlockConfiguration');
      node.addPropertyDeletionOverride('OwnershipControls');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('InsightSelectors');
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
    if (node instanceof cdk.aws_ec2.CfnVPCEndpoint) {
      const ServiceName = node.serviceName.replace('com.amazonaws.us', 'gov.ic.c2s.us');
      node.addPropertyOverride('ServiceName', ServiceName);
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = (node as cdk.aws_iam.CfnRole).assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc.toJSON()).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.c2shome.ic.gov/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for CN environments
 */
class CnOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
  }
}

/**
 * Default memory override for Lambda resources
 */
class LambdaDefaultMemoryAspect implements cdk.IAspect {
  visit(node: IConstruct): void {
    if (node instanceof cdk.CfnResource) {
      if (node.cfnResourceType === 'AWS::Lambda::Function') {
        const memorySize = (node as cdk.aws_lambda.CfnFunction).memorySize;
        if (!memorySize || memorySize < 256) {
          node.addPropertyOverride('MemorySize', 256);
        }
      }
    }
  }
}

/**
 * Solution ID override for Lambda resources
 */
class AwsSolutionAspect implements cdk.IAspect {
  visit(node: IConstruct): void {
    if (node instanceof cdk.CfnResource) {
      if (node.cfnResourceType === 'AWS::Lambda::Function') {
        node.addPropertyOverride('Environment.Variables.SOLUTION_ID', `AwsSolution/SO0199/${version}`);
      }
    }
  }
}

/**
 * Add accelerator specific aspects to the application based on partition
 */
export class AcceleratorAspects {
  /**
   * The region for global API endpoints
   * based on AWS partition
   */
  public readonly globalRegion: string;

  constructor(app: cdk.App, partition: string) {
    let globalRegion = 'us-east-1';
    // Add partition specific overrides
    switch (partition) {
      case 'aws-us-gov':
        globalRegion = 'us-gov-west-1';
        cdk.Aspects.of(app).add(new GovCloudOverrides());
        break;
      case 'aws-iso':
        globalRegion = 'us-iso-east-1';
        cdk.Aspects.of(app).add(new IsoOverrides());
        break;
      case 'aws-iso-b':
        globalRegion = 'us-isob-east-1';
        cdk.Aspects.of(app).add(new IsobOverrides());
        break;
      case 'aws-cn':
        globalRegion = 'cn-northwest-1';
        cdk.Aspects.of(app).add(new CnOverrides());
        break;
    }
    // Add default aspects
    cdk.Aspects.of(app).add(new LambdaDefaultMemoryAspect());
    cdk.Aspects.of(app).add(new AwsSolutionAspect());

    // Set global region
    this.globalRegion = globalRegion;
  }
}
