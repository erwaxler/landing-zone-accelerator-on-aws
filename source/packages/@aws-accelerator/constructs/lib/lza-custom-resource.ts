/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NagSuppressions } from 'cdk-nag';
import { v4 as uuidv4 } from 'uuid';
import { pascalCase } from 'change-case';

/**
 * Initialized LzaCustomResourceProps properties
 */
export interface LzaCustomResourceProps {
  /**
   * Custom resource properties
   */
  readonly resource: {
    /**
     * Logical name for the custom resource
     */
    readonly name: string;
    /**
     * Logical Id for the implementor construct
     */
    readonly parentId: string;
    /**
     * A name value object list for custom resource properties
     *
     * @example
     * [{ solution: 'accelerator', author: 'tsd' }]
     */
    readonly properties?: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      [key: string]: any;
    }[];
    /**
     * Generates a new UUID to force the resource to update
     */
    readonly forceUpdate?: boolean;
  };

  /**
   * Custom resource lambda properties
   */
  readonly lambda: {
    /**
     * LZA Custom resource lambda asset folder path including the /dist folder
     */
    readonly assetPath: string;
    /**
     * Custom resource lambda environment encryption key
     */
    readonly environmentEncryptionKmsKey: cdk.aws_kms.IKey;
    /**
     * Custom resource lambda log group encryption key
     */
    readonly cloudWatchLogKmsKey: cdk.aws_kms.IKey;
    /**
     * Custom resource lambda cloudwatch log retention in days
     */
    readonly cloudWatchLogRetentionInDays: number;
    /**
     * A description of the custom resource lambda function.
     */
    readonly description?: string;
    /**
     * Custom resource lambda execution role. This is the role that will be assumed by the function upon execution.
     * @default - A unique role will be generated for this lambda function.
     */
    readonly role?: cdk.aws_iam.Role;
    /**
     * The amount of memory, in MB, that is allocated to custom resource Lambda function. Lambda uses this value to proportionally allocate the amount of CPU power.
     * @default 256
     */
    readonly memorySize?: number;
    /**
     * Custom resource lambda function execution time (in seconds) after which Lambda terminates the function.
     * @default 3 seconds
     */
    readonly timeOut?: number;
    /**
     * Initial policy statements to add to the created custom resource Lambda Role.
     */
    readonly roleInitialPolicy?: cdk.aws_iam.PolicyStatement[];
    /**
     * The name of the method within lambda code that custom resource lambda calls to execute the function. The format includes the file name.
     * @default 'index.handler'
     */
    readonly handler?: string;
    /**
     * Determine the removal policy of CloudWatch log group for the lambda.
    @default RemovalPolicy.Destroy
     */
    readonly cloudWatchLogRemovalPolicy?: cdk.RemovalPolicy;
    /**
     * Debug flag used in custom resource lambda function to output debug logs
    @default false    
     */
    readonly debug?: boolean;
    /**
     * A name value object list for lambda environment variables
     *
     * @example
     * [{ solution: 'accelerator', author: 'tsd' }]
     */
    readonly environmentVariables?: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      [key: string]: any;
    }[];
  };
}

/**
 * Class for LZA Custom Resource Construct
 * This class can create LZA standard custom resource constructs
 */
export class LzaCustomResource extends Construct {
  public readonly resource: cdk.CustomResource;

  private readonly provider: cdk.custom_resources.Provider;

  constructor(scope: Construct, id: string, props: LzaCustomResourceProps) {
    super(scope, id);

    const functionName = pascalCase(props.resource.name + 'Function');

    const providerLambdaFunction = new cdk.aws_lambda.Function(this, functionName, {
      description:
        props.lambda.description ?? `Accelerator deployed ${props.resource.name} custom resource lambda function.`,
      code: cdk.aws_lambda.Code.fromAsset(props.lambda.assetPath),
      runtime: cdk.aws_lambda.Runtime.NODEJS_16_X,
      memorySize: props.lambda.memorySize ?? 256,
      timeout: props.lambda.timeOut ? cdk.Duration.seconds(props.lambda.timeOut) : cdk.Duration.seconds(160),
      role: props.lambda.role,
      initialPolicy: props.lambda.roleInitialPolicy,
      handler: props.lambda.handler ?? 'index.handler',
      environmentEncryption: props.lambda.environmentEncryptionKmsKey,
      environment: this.prepareLambdaEnvironments(props),
    });

    new cdk.aws_logs.LogGroup(this, `${providerLambdaFunction.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${providerLambdaFunction.functionName}`,
      retention: props.lambda.cloudWatchLogRetentionInDays,
      encryptionKey: props.lambda.cloudWatchLogKmsKey,
      removalPolicy: props.lambda.cloudWatchLogRemovalPolicy ?? cdk.RemovalPolicy.DESTROY,
    });

    this.provider = new cdk.custom_resources.Provider(this, 'Resource', {
      onEventHandler: providerLambdaFunction,
    });

    this.resource = new cdk.CustomResource(this, pascalCase(props.resource.name + 'Resource'), {
      serviceToken: this.provider.serviceToken,
      properties: this.prepareResourceProperties(props),
    });

    const stack = cdk.Stack.of(scope);

    // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${props.resource.parentId}/${id}/${functionName}/ServiceRole/Resource`,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'AWS Custom resource provider framework-role created by cdk.',
        },
      ],
    );

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${props.resource.parentId}/${id}/${functionName}/ServiceRole/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Allows only specific policy.',
        },
      ],
    );

    // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${props.resource.parentId}/${id}/Resource/framework-onEvent/ServiceRole/Resource`,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'AWS Custom resource provider framework-role created by cdk.',
        },
      ],
    );

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${props.resource.parentId}/${id}/Resource/framework-onEvent/ServiceRole/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Allows only specific policy.',
        },
      ],
    );
  }

  /**
   * Function to prepare Lambda Environment variables
   * @param props {@link LzaCustomResourceProps}
   * @returns
   */
  private prepareLambdaEnvironments(props: LzaCustomResourceProps):
    | {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        [key: string]: any;
      }
    | undefined {
    const lambdaEnvironmentList:
      | {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          [key: string]: any;
        }[] = props.lambda.environmentVariables ?? [];

    const lambdaEnvironment: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      [key: string]: any;
    } = {};

    for (const environmentVariable of lambdaEnvironmentList) {
      for (const [key, value] of Object.entries(environmentVariable)) {
        lambdaEnvironment[key] = value;
      }
    }

    return lambdaEnvironmentList.length > 0 ? lambdaEnvironment : undefined;

    // return lambdaEnvironment;
  }

  /**
   * Function to prepare Resource Properties
   * @param props {@link LzaCustomResourceProps}
   * @returns
   */
  private prepareResourceProperties(props: LzaCustomResourceProps):
    | {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        [key: string]: any;
      }
    | undefined {
    const resourcePropertyList:
      | {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          [key: string]: any;
        }[] = props.resource.properties ?? [{ debug: props.lambda.debug ?? false }];

    if (props.resource.forceUpdate) {
      resourcePropertyList.push({ uuid: uuidv4() });
    }

    const resourceProperties: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      [key: string]: any;
    } = {};

    for (const resourceProperty of resourcePropertyList) {
      for (const [key, value] of Object.entries(resourceProperty)) {
        resourceProperties[key] = value;
      }
    }

    return resourcePropertyList.length > 0 ? resourceProperties : undefined;
  }
}
