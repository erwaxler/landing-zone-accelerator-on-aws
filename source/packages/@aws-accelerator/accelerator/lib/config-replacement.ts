/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as fs from 'fs';
import * as path from 'path';

import {
  AccountsConfig,
  CustomizationsConfig,
  GlobalConfig,
  IamConfig,
  NetworkConfig,
  OrganizationConfig,
  ReplacementsConfig,
  SecurityConfig,
} from '@aws-accelerator/config';
import { createLogger } from '@aws-accelerator/utils/lib/logger';
import { Accelerator, shouldLookupDynamoDb } from './accelerator';

const logger = createLogger(['config-replacement']);
const configDirPath = process.argv[2];
const homeRegion = GlobalConfig.loadRawGlobalConfig(configDirPath).homeRegion;

const fileNameList = [
  CustomizationsConfig.FILENAME,
  GlobalConfig.FILENAME,
  IamConfig.FILENAME,
  NetworkConfig.FILENAME,
  OrganizationConfig.FILENAME,
  SecurityConfig.FILENAME,
];

const enableSingleAccountMode = process.env['ACCELERATOR_ENABLE_SINGLE_ACCOUNT_MODE']
  ? process.env['ACCELERATOR_ENABLE_SINGLE_ACCOUNT_MODE'] === 'true'
  : false;

const props = {
  partition: process.env['PARTITION'] ?? 'aws',
  region: process.env['AWS_REGION'],
  account: process.env['ACCOUNT_ID'],
  enableSingleAccountMode: enableSingleAccountMode,
  // stage is needed to find if dynamoDB lookup is needed. ACCELERATOR_STAGE is set in codePipeline environments
  stage: process.env['ACCELERATOR_STAGE'],
};

if (configDirPath) {
  logger.info(`Config source directory - ${configDirPath}`);
  processReplacements(props);
} else {
  logger.error('Config source directory undefined !!!');
  process.exit(1);
}

async function processReplacements(props: {
  partition: string;
  region: string | undefined;
  enableSingleAccountMode: boolean;
  account: string | undefined;
  stage: string | undefined;
}) {
  await Accelerator.getManagementAccountCredentials(props.partition);
  const orgsEnabled = OrganizationConfig.loadRawOrganizationsConfig(configDirPath).enable;
  const loadFromDynamoDbTable = shouldLookupDynamoDb(props.stage);

  // Load accounts config
  let accountsConfig: AccountsConfig | undefined = undefined;
  try {
    accountsConfig = AccountsConfig.load(configDirPath);
    await accountsConfig.loadAccountIds(
      props.partition,
      props.enableSingleAccountMode,
      orgsEnabled,
      accountsConfig,
      undefined,
      loadFromDynamoDbTable,
    );
  } catch (e) {
    logger.error(`Error loading accounts config: ${e}`);
    process.exit(1);
  }

  // Load replacements config
  let replacementsConfig: ReplacementsConfig | undefined = undefined;
  try {
    replacementsConfig = ReplacementsConfig.load(configDirPath, accountsConfig);
    await replacementsConfig.loadDynamicReplacements(homeRegion);
  } catch (e) {
    logger.error(`Error loading replacements config: ${e}`);
    process.exit(1);
  }

  // Process each config file
  for (const fileName of fileNameList) {
    const filePath = path.join(configDirPath, fileName);

    // Skip if file doesn't exist
    if (!fs.existsSync(filePath)) {
      logger.info(`Skipping ${fileName} as it does not exist`);
      continue;
    }

    try {
      // Read the original file
      const originalContent = fs.readFileSync(filePath, 'utf8');

      // Process replacements
      const processedContent = replacementsConfig.preProcessBuffer(originalContent);

      // Create output file with -replaced suffix
      const outputFilePath = filePath.replace('.yaml', '-replaced.yaml');
      fs.writeFileSync(outputFilePath, processedContent);

      logger.info(`Successfully processed ${fileName} -> ${path.basename(outputFilePath)}`);
    } catch (e) {
      logger.error(`Error processing ${fileName}: ${e}`);
    }
  }

  logger.info('Configuration replacement completed successfully');
}
