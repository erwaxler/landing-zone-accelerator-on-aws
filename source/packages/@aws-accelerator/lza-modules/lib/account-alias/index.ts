/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import {
  IAMClient,
  ListAccountAliasesCommand,
  CreateAccountAliasCommand,
  DeleteAccountAliasCommand,
  ListAccountAliasesCommandOutput,
} from '@aws-sdk/client-iam';
import { getGlobalRegion } from '@aws-accelerator/utils/lib/common-functions';
import { getAccountAliasesFromConfig, getManagementAccountCredentials } from '../../common/functions';

import { createLogger } from '@aws-accelerator/utils/lib/logger';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';

import { AcceleratorModule } from '../accelerator-module';
import { ModuleOptionsType } from '../../common/resources';
import { AccountsConfig, AccountConfig, GlobalConfig } from '@aws-accelerator/config';
import * as winston from 'winston';
import path from 'path';

/**
 * AccountAlias class to manage Account Alias operations
 */
export class AccountAlias implements AcceleratorModule {
  private readonly logger: winston.Logger = createLogger([path.parse(path.basename(__filename)).name]);
  /**
   * Handler function to manage Account Aliases
   *
   * @remarks
   * The following activities are performed by this function for each account
   * - Check if a new account alias is needed
   * - Delete existing account alias if one exists
   * - Create new account alias
   *
   * @param module string
   * @param props {@link ModuleOptionsType}
   * @returns status string
   */
  public async handler(module: string, props: ModuleOptionsType): Promise<string> {
    const statuses: string[] = [];
    // Get accounts from accounts config
    const accountsConfig = AccountsConfig.load(props.configDirPath);
    const enableSingleAccountMode = process.env['ACCELERATOR_ENABLE_SINGLE_ACCOUNT_MODE']
      ? process.env['ACCELERATOR_ENABLE_SINGLE_ACCOUNT_MODE'] === 'true'
      : false;

    const accounts = accountsConfig.getAccounts(enableSingleAccountMode);

    const accountAliases = await getAccountAliasesFromConfig(accountsConfig, enableSingleAccountMode);
    if (accountAliases.length! > 0) {
      return `No aliases in accounts config, module "${module}" execution skipped`;
    }

    const globalRegion = getGlobalRegion(props.partition);

    const globalConfig = GlobalConfig.load(props.configDirPath);
    const managementAccountCredentials = await getManagementAccountCredentials(
      props.partition,
      globalConfig.homeRegion,
      props.solutionId,
    );

    const iamClient = new IAMClient({
      region: globalRegion,
      customUserAgent: props.solutionId,
      retryStrategy: setRetryStrategy(),
      credentials: managementAccountCredentials,
    });

    accounts.forEach(account => {
      if (!account.accountAlias) {
        this.logger.info(`Account alias not configured for account ${account.name}`);
      } else {
        try {
          this.manageAccountAlias(iamClient, account, statuses);
        } catch (e) {
          this.logger.error(`Error while managing account alias for account ${account.name}`);
          this.logger.error(e);
          throw e;
        }
      }
    });

    return `Module "${module}" completed with following status.\n ${statuses.join('\n')}`;
  }

  /**
   * Function to manage the account alias of an account
   * @param iamClient {@link IAMClient}
   * @param account {@link AccountConfig}
   * @param statuses string[]
   */
  private async manageAccountAlias(iamClient: IAMClient, account: AccountConfig, statuses: string[]): Promise<void> {
    const localStatuses: string[] = [];

    // Get current account alias
    let listAliasResponse: ListAccountAliasesCommandOutput;
    try {
      listAliasResponse = await throttlingBackOff(() => iamClient.send(new ListAccountAliasesCommand({})));
    } catch (error) {
      throw new Error(`Failed to list account aliases: ${error}`);
    }

    // There can never be more than one account alias https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListAccountAliases.html#API_ListAccountAliases_ResponseElements
    const currentAlias = listAliasResponse.AccountAliases?.[0];

    // If current alias matches desired alias, no action needed
    if (currentAlias === account.accountAlias) {
      this.logger.info(`Account alias "${account.accountAlias}" already set`);
      return;
    }

    // Delete existing alias if one exists
    if (currentAlias) {
      try {
        await throttlingBackOff(() =>
          iamClient.send(
            new DeleteAccountAliasCommand({
              AccountAlias: currentAlias,
            }),
          ),
        );
        localStatuses.push(`Successfully deleted existing account alias "${currentAlias}"`);
      } catch (error) {
        throw new Error(`Failed to delete existing alias "${currentAlias}": ${error}`);
      }
    }

    // Create new alias
    try {
      await throttlingBackOff(() =>
        iamClient.send(
          new CreateAccountAliasCommand({
            AccountAlias: account.accountAlias,
          }),
        ),
      );
      localStatuses.push(`Account alias "${account.accountAlias}" successfully set for account "${account.name}".`);
    } catch (error) {
      this.logger.error(`Account "${account.name}" may be left without an alias`);
      throw new Error(`Failed to create new alias "${account.accountAlias}": ${error}`);
    }

    statuses.push(...localStatuses);
  }
}
