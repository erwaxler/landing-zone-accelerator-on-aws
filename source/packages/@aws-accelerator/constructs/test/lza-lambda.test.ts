/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { LzaLambda } from '../lib/lza-lambda';
import { snapShotTest } from './snapshot-test';
import { describe } from '@jest/globals';
const testNamePrefix = 'Construct(IdentityCenterGetInstanceId): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new LzaLambda(stack, 'LzaLambda', {
  functionName: 'LzaLambda',
  assetPath: `${__dirname}/../lib/aws-identitystore/get-users-groups-id/dist`,
  environmentEncryptionKmsKey: new cdk.aws_kms.Key(stack, 'LambdaKey', {}),
  cloudWatchLogKmsKey: new cdk.aws_kms.Key(stack, 'CloudWatchKey', {}),
  cloudWatchLogRetentionInDays: 365,
  description: 'LZA Lambda Function for Snapshot test',
  timeOut: cdk.Duration.minutes(5),
  roleInitialPolicy: [
    new cdk.aws_iam.PolicyStatement({
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['identitystore:ListGroups', 'identitystore:ListUsers'],
      resources: ['*'],
    }),
  ],
  cloudWatchLogRemovalPolicy: cdk.RemovalPolicy.RETAIN,
  nagSuppressionPrefix: `testPrefix`,
  environmentVariables: [{ envKey: 'envValue' }],
});

/**
 * LzaLambda construct test
 */
describe('LzaLambda', () => {
  snapShotTest(testNamePrefix, stack);
});
