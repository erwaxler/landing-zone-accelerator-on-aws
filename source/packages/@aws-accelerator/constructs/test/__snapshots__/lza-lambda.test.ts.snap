// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LzaLambda Construct(IdentityCenterGetInstanceId):  Snapshot Test 1`] = `
{
  "Resources": {
    "CloudWatchKey9B181885": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "LambdaKey984A39D9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "LzaLambdaF8E06A14": {
      "DependsOn": [
        "LzaLambdaServiceRoleDefaultPolicy88664093",
        "LzaLambdaServiceRole1136AAD1",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "LZA Lambda Function for Snapshot test",
        "Environment": {
          "Variables": {
            "envKey": "envValue",
          },
        },
        "FunctionName": "LzaLambda",
        "Handler": "index.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "LambdaKey984A39D9",
            "Arn",
          ],
        },
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "LzaLambdaServiceRole1136AAD1",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LzaLambdaResourceLogGroup2399FC9C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CloudWatchKey9B181885",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LzaLambdaF8E06A14",
              },
            ],
          ],
        },
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "LzaLambdaServiceRole1136AAD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LzaLambdaServiceRoleDefaultPolicy88664093": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "identitystore:ListGroups",
                "identitystore:ListUsers",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LzaLambdaServiceRoleDefaultPolicy88664093",
        "Roles": [
          {
            "Ref": "LzaLambdaServiceRole1136AAD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
